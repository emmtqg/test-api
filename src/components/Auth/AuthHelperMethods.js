import decode from 'jwt-decode';

import config from '../Config';

// The apiCamelizedLookup provides a fieldname check and crossreference
// for the form field names. It aids in translation of valid property
// names to whatever the API required (single point of definition between
// the state properties and api names). It was assumed if the ESLint rules
// did not allow underscores (as specified in the API names) then, rather
// than overriding the linter, a mechanism should be incorporated to allow
// any API to specify properties as required and they can then be updated
// in one location. If form fields are checked for testing, then the form
// component can also utilize the apiCamelizedLookup for field specifications.
const { apiCamelizedLookup } = config;
const apiNames = Object.keys(apiCamelizedLookup);

export default class AuthHelperMethods {
  getApiPropName = (name) => {
    // Get the API 'expected' property of input form field
    let index = -1;
    Object.values(apiCamelizedLookup).findIndex((ele, i) => {
      index = i;
      return (ele === name);
    });
    return index; // returning matched index or -1 if not in apiCamelized object
  }

  lookupApiName = (name) => {
    const index = this.getApiPropName(name);
    if (index > -1) {
      // apiNames are the API keys in lookup array
      return apiNames[index];
    }

    return ''; // API will error with invalid fieldname
  }

  register = (registerObject) => {
    const {
      firstName,
      lastName,
      email,
      password,
      confirmPassword,
    } = registerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.registerUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          email,
          password,
          [this.lookupApiName('confirmPassword')]: confirmPassword,
        }),
      })
        .then((response) => {
          // console.log(`Auth (Register): ${response}`);
          if (response) {
            this.setToken(response.token); // Setting the token in localStorage
            return Promise.resolve(response);
          }
          return Promise.reject(response);
        })
    );
  }

  newPlayer = (newPlayerObject) => {
    const {
      firstName,
      lastName,
      rating,
      handedness,
    } = newPlayerObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.playersUrl, {
        method: 'POST',
        body: JSON.stringify({
          [this.lookupApiName('firstName')]: firstName,
          [this.lookupApiName('lastName')]: lastName,
          rating,
          handedness,
        }),
      })
    );
  }

  getPlayers = () => (
    this.fetch(config.playersUrl, {
      method: 'GET',
    })
  );

  removePlayer = (playerId) => {
    const deleteUrl = `${config.playersUrl}/${playerId}`;
    return (
      this.fetch(deleteUrl, {
        method: 'DELETE',
      })
    );
  }

  login = (loginObject) => {
    const {
      email,
      password,
    } = loginObject;

    // Get a token from api server using the fetch api
    return (
      this.fetch(config.loginUrl, {
        method: 'POST',
        body: JSON.stringify({
          email,
          password,
        }),
      })
        .then((response) => {
          if (response) {
            // console.log(`Auth (Login): ${response}`);
            this.setToken(response.token); // Setting the token in localStorage
            return response;
          }
          return Promise.reject(response);
        })
        .catch((error) => {
          if (error && error.message) {
            // console.log(`Error: ${error.message}`);
          }
          return Promise.reject(error);
        })
    );
  }

  loggedIn = () => {
    // Checks if there is a saved token and it's still valid
    const token = this.getToken(); // Getting token from localstorage
    return (typeof token !== typeof undefined) && (!!token && !this.isTokenExpired(token));
  }

  isTokenExpired = (token) => {
    try {
      const decoded = decode(token);
      if (decoded.exp < Date.now() / 1000) { // Checking if token is expired.
        return true;
      }
      return false;
    } catch (err) {
      // console.log('Check for expired tolen failed.');
      return false;
    }
  }

  setToken = (idToken) => {
    // Saves user token to localStorage
    localStorage.setItem('id_token', idToken);
  }

  getToken = () =>
    // Retrieves the user token from localStorage
    localStorage.getItem('id_token')


  logout = () => {
    // Clear user token and profile data from localStorage
    localStorage.removeItem('id_token');
  }

  getConfirm = () => {
    // Using jwt-decode npm package to decode the token
    const answer = decode(this.getToken());
    // console.log(`Get confirm token => ${answer}`);
    return answer;
  }

  fetch = (url, options) => {
    // performs api calls sending the required authentication headers
    const headers = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    };

    // Setting Authorization header
    // Authorization: Bearer xxxxxxx.xxxxxxxx.xxxxxx
    if (this.loggedIn()) {
      headers.Authorization = `Bearer ${this.getToken()}`;
    }

    return fetch(url, {
      headers,
      ...options,
    })
      .then(this.checkStatus)
      .then(response => response.json())
      .catch((error) => {
        // console.log(`Error: ${error.message}`);
        throw error;
      });
  }

  checkStatus = (response) => {
    // Success status lies between 200 to 300
    if (response.status >= 200 && response.status < 300) {
      return response;
    }

    // raises an error in case response status is not a success
    const error = new Error(response.statusText);
    error.response = response;
    error.status = response.status;
    throw error;
  }
}
