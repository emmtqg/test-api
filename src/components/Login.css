.center {
  font-family: 'Raleway', sans-serif;
  background-color: #1ABC9C;
  position: absolute;
  height: 100%;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}
