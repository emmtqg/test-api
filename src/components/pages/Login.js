import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import FormControl from '@material-ui/core/FormControl';
import Input from '@material-ui/core/Input';
import InputLabel from '@material-ui/core/InputLabel';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import withStyles from '@material-ui/core/styles/withStyles';

const styles = theme => ({
  main: {
    maxWidth: 750,
    width: 'auto',
    display: 'block', // Fix IE 11 issue.
    [theme.breakpoints.down('sm')]: {
      marginLeft: 5,
      marginRight: 5,
    },
    [theme.breakpoints.up((400 + theme.spacing.unit) * 3 * 2)]: {
      width: 400,
      marginLeft: 'auto',
      marginRight: 'auto',
    },
  },
  paper: {
    marginTop: theme.spacing.unit * 8,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 3}px ${theme.spacing.unit * 3}px`,
  },
  avatar: {
    margin: theme.spacing.unit,
    backgroundColor: theme.palette.secondary.main,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing.unit,
  },
  submit: {
    marginTop: theme.spacing.unit * 3,
  },
});

class Login extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      password: '',
      remember: false,
      message: '',
      messageSuggest: '',
    };

    this.handleFormSubmit = this.handleFormSubmit.bind(this);
  }

  componentWillMount() {
    // Here is a great place to redirect someone who is already logged in
    if (this.props.auth.loggedIn()) {
      this.props.history.replace('/');
    }
  }

  handleChange = (event) => {
    this.setState({ [event.target.name]: event.target.value });
  }

  handleFormSubmit = (e) => {
    e.preventDefault();

    // Here is where all the login logic will go. Upon clicking the login button,
    // we would like to utilize a login method that will send our entered
    // credentials over to the server for verification. Once verified, it should
    // store your token and send you to the protected route.

    this.props.auth.login(this.state)
      .then((response) => {
        // JWT is returned in response json
        if (typeof response.token === typeof undefined) {
          this.setState({
            message: 'Could not set your security token.',
            messageSuggest: 'Please try logging in again.',
          });
        }
        this.setState({
          message: 'Successful Login!',
          messageSuggest: '',
        });
        return this.props.history.replace('/roster');
      })
      .catch((error) => {
        this.setState({
          message: `Login encountered an error: (${error.status}) ${error.message}.`,
          messageSuggest: 'Please recheck your password.',
        });
      });

    return null;
  }

  render() {
    const { classes } = this.props;
    return (
      <main className={classes.main}>
        <CssBaseline />
        <Paper className={classes.paper}>
          <Avatar className={classes.avatar}>
            <LockOutlinedIcon />
          </Avatar>
          <Typography component="h1" variant="h5">
            Log In
          </Typography>
          { (this.state.message) ? <div><br /><hr /></div> : null }
          <Typography style={{ color: 'red' }} component="h6" variant="h6">
            {this.state.message}<br />
            {this.state.messageSuggest}
          </Typography>
          <form id="formLogin" className={classes.form} onSubmit={this.handleFormSubmit}>
            <FormControl margin="normal" required fullWidth>
              <InputLabel htmlFor="email">Email Address</InputLabel>
              <Input id="email" name="email" autoComplete="email" autoFocus onChange={this.handleChange} />
            </FormControl>
            <FormControl margin="normal" required fullWidth>
              <InputLabel htmlFor="password">Password</InputLabel>
              <Input name="password" type="password" id="password" autoComplete="current-password" onChange={this.handleChange} />
            </FormControl>
            <Button
              type="submit"
              id="login"
              fullWidth
              variant="contained"
              color="primary"
              className={classes.submit}
              onSubmit={this.handleSubmit}
            >
              Sign in
            </Button>
          </form>
        </Paper>
      </main>
    );
  }
}

Login.defaultProps = {
  auth: {},
  // History object passed via Router or a function
  // that passed in unit tests via jest.
  history: PropTypes.oneOfType([
    PropTypes.shape({
      replace: () => {},
    }),
    () => {},
  ]),
};

Login.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  classes: PropTypes.object.isRequired,
  auth: PropTypes.shape({
    loggedIn: PropTypes.func,
    login: PropTypes.func,
  }),
  // History object passed via Router or a function
  // that passed in unit tests via jest.
  history: PropTypes.oneOfType([
    PropTypes.shape({
      replace: PropTypes.func,
    }),
    PropTypes.func,
  ]),
};

export default withStyles(styles)(Login);
