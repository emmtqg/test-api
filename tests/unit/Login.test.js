import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import Login from '../../src/components/pages/Login';
import AuthHelperMethods from '../../src/components/auth/AuthHelperMethods';

let UnwrappedShallowComponent;
let wrapper;

const auth = new AuthHelperMethods();

// auth mocks
auth.loggedIn = jest.fn(() => false); // replace loggedIn test for component mount check
auth.setToken = jest.fn(() => {
  auth.setToken('eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI1YzlhNzFiMDMxMGQyMTdmZjM5MDFlMjkiLCJpYXQiOjE1NTM3MTQzNTd9.sfyRdYlfXXzgqFfxLuEONxh1TMG1teWTAMO_GvG4Xdw');
});

// auth mocks for Promise return of newLogin API call from AuthHelpers.js
const newLogin = {
  success: true,
  loginResponse: {
    id: '5b24deb1310d217ff39012c4',
    token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiI1YzlhNzFiMDMxMGQyMTdmZjM5MDFlMjkiLCJpYXQiOjE1NTM3MTQzNTd9.sfyRdYlfXXzgqFfxLuEONxh1TMG1teWTAMO_GvG4Xdw',
  },
};
const loginMock = jest.fn().mockReturnValue(Promise.resolve(newLogin));
auth.login = loginMock;

// history mock
const history = {
  replace: jest.fn(),
};

// Although no login form fields need to be camelized for the state properties,
// the object serves to list the valid fields for the form.
const apiCamelizedLookupForLogin = {
  email: 'email',
  password: 'password',
};

beforeAll(() => {
  wrapper = mount(<Login auth={auth} props={history} />);
  const shallowComponent = shallow(<Login auth={auth} props={history} />);
  UnwrappedShallowComponent = shallowComponent.first().shallow();
});

afterAll(() => {
  // wrapper.unmount();
});

describe('The Login Form should', () => {
  it('match the Login form snapshot', () => {
    expect(wrapper.find('Login').length).toEqual(1);
  });

  it('render correctly', () => {
    const component = renderer.create(<Login auth={auth} props={history} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('actually have a form element', () => {
    expect(wrapper.find('form').length).toBe(1);
  });

  describe('The form fields retrieve data and update component state', () => {
    const formNames = Object.keys(apiCamelizedLookupForLogin);
    let input;
    formNames.forEach((element) => {
      it(` checking ${element} has a defined onChange function`, () => {
        // MUI returning multiple nodes with #text_field, get the first with props
        input = wrapper.find(`#${element}`).first();
        expect(input.props('onChange')).toBeDefined();
        expect(typeof input.props('onChange').onChange === 'function').toBe(true);
      });
    });

    formNames.forEach((element) => {
      it(` checking ${element} updates state property on change`, () => {
        input = UnwrappedShallowComponent.find(`#${element}`).first();
        input.simulate('change', { target: { name: element, value: 'Ping' } });
        expect(UnwrappedShallowComponent.state([element])).toBe('Ping');
      });
    });
  });

  describe('The Auth module (mock) should be called to log in user on Submit', () => {
    it('should call handleSubmit and mock Auth API call to login and return Promise', () => {
      // mock history.replace('/roster')
      const mockReplaceURLtoRoster = jest.fn(() => true);
      const mockEvent = { preventDefault: () => {} };
      history.replace = mockReplaceURLtoRoster;

      const wrap = shallow(<Login auth={auth} props={history} />);
      const unwrappedWrap = wrap.first().shallow();
      let input;

      input = unwrappedWrap.find('#email').first();
      input.simulate('change', { target: { email: 'email', value: 'sue@magoo.com' } });
      input = unwrappedWrap.find('#password').first();
      input.simulate('change', { target: { password: 'password', value: 'a' } });
      input = unwrappedWrap.find('#password').first();

      // Call submit and mock Auth.login to return resolved promise
      unwrappedWrap.find('form').simulate('submit', mockEvent);
      expect(loginMock('auth.login')).resolves.toBe(newLogin);
    });
  });
});
