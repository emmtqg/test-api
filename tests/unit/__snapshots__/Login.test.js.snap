// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Login Form should render correctly 1`] = `
<main
  className="Login-main-1"
>
  <div
    className="MuiPaper-root-6 MuiPaper-elevation2-10 MuiPaper-rounded-7 Login-paper-2"
  >
    <div
      className="MuiAvatar-root-33 MuiAvatar-colorDefault-34 Login-avatar-3"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-36"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <g
          fill="none"
        >
          <path
            d="M0 0h24v24H0V0z"
          />
          <path
            d="M0 0h24v24H0V0z"
            opacity=".87"
          />
        </g>
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
        />
      </svg>
    </div>
    <h1
      className="MuiTypography-root-45 MuiTypography-h5-61"
    >
      Log In
    </h1>
    <h6
      className="MuiTypography-root-45 MuiTypography-h6-62"
      style={
        Object {
          "color": "red",
        }
      }
    >
      
      <br />
      
    </h6>
    <form
      className="Login-form-4"
      id="formLogin"
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root-81 MuiFormControl-marginNormal-82 MuiFormControl-fullWidth-84"
      >
        <label
          className="MuiFormLabel-root-96 MuiFormLabel-required-101 MuiInputLabel-required-89 MuiInputLabel-root-85 MuiInputLabel-formControl-90 MuiInputLabel-animated-93"
          data-shrink={false}
          htmlFor="email"
        >
          Email Address
          <span
            className="MuiFormLabel-asterisk-102"
          >
             *
          </span>
        </label>
        <div
          className="MuiInputBase-root-116 MuiInput-root-103 MuiInput-underline-107 MuiInputBase-formControl-117 MuiInput-formControl-104"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="email"
            autoFocus={true}
            className="MuiInputBase-input-126 MuiInput-input-111"
            disabled={false}
            id="email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-81 MuiFormControl-marginNormal-82 MuiFormControl-fullWidth-84"
      >
        <label
          className="MuiFormLabel-root-96 MuiFormLabel-required-101 MuiInputLabel-required-89 MuiInputLabel-root-85 MuiInputLabel-formControl-90 MuiInputLabel-animated-93"
          data-shrink={false}
          htmlFor="password"
        >
          Password
          <span
            className="MuiFormLabel-asterisk-102"
          >
             *
          </span>
        </label>
        <div
          className="MuiInputBase-root-116 MuiInput-root-103 MuiInput-underline-107 MuiInputBase-formControl-117 MuiInput-formControl-104"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="current-password"
            className="MuiInputBase-input-126 MuiInput-input-111 MuiInputBase-inputType-129 MuiInput-inputType-114"
            disabled={false}
            id="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root-159 MuiButton-root-133 MuiButton-contained-144 MuiButton-containedPrimary-145 MuiButton-raised-147 MuiButton-raisedPrimary-148 MuiButton-fullWidth-158 Login-submit-5"
        disabled={false}
        id="login"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="submit"
      >
        <span
          className="MuiButton-label-134"
        >
          Sign in
        </span>
        <span
          className="MuiTouchRipple-root-162"
        />
      </button>
    </form>
  </div>
</main>
`;
